Update models docs to core 2.22.0 (#196)
OCR-D/ocrd-website: site/en/models.md

at a specific location (tesseract, ocropy, kraken) while others require the full
path to a model (calamari).

Since [v2.22.0](https://github.com/OCR-D/core/releases/v2.22.0), OCR-D/core
comes with a framework for managing processor resources uniformly. This means

> that OCR-D/core will take care of lookin in well-defined places in the
> filesystem for resources for specific processors. It also knows how to cache

Reviewer: ```suggestion
that processors can delegate to OCR-D/core to resolve specific file resources by name,
looking in well-defined places in the filesystem. This also includes downloading and caching
```

> file parameters passed as a URL. OCR-D/core also comes with a bundled database

Reviewer 2: ```suggestion
file parameters passed as a URL. Furthermore, OCR-D/core comes with a bundled database
```

> of known resources, such as OCR models, configurations and other
> processor-specific data. This means that OCR-D users should be able to

Reviewer: ```suggestion
of known resources, such as models, dictionaries, configurations and other
processor-specific data files. This means that OCR-D users should be able to
```

concentrate on fine-tuning their OCR workflows and not bother with implementation
details like "where do I get models from and where do I put them".

> 

Reviewer: ```suggestion
In particular, users can reference file parameters by name now.

```

All of the above mentioned functionality can be accessed using the `ocrd
resmgr` command line tool.

## What models are available?

To get a list of the resources that the OCR-D/core [is aware
of](https://github.com/OCR-D/core/blob/master/ocrd/ocrd/resource_list.yml):

```
ocrd resmgr list-available
```

The output will look similar to this:

```

ocrd-calamari-recognize
- qurator-gt4hist-0.3 (https://qurator-data.de/calamari-models/GT4HistOCR/2019-07-22T15_49+0200/model.tar.xz)
  Calamari model trained with GT4HistOCR
- qurator-gt4hist-1.0 (https://qurator-data.de/calamari-models/GT4HistOCR/2019-12-11T11_10+0100/model.tar.xz)
  Calamari model trained with GT4HistOCR

ocrd-cis-ocropy-recognize
- LatinHist.pyrnn.gz (https://github.com/chreul/OCR_Testdata_EarlyPrintedBooks/raw/master/LatinHist-98000.pyrnn.gz)
  ocropy historical latin model by github.com/chreul
```

> As you can see, resources are grouped by the processor they are used by.

Reviewer: ```suggestion
As you can see, resources are grouped by the processors which make use of them.
```

The word after the list symbol, e.g. `qurator-gt4hist-0.3`,

> `LatinHist.pyrnn.gz`, define the "name" of the resource, a shorthand you can

Reviewer: ```suggestion
`LatinHist.pyrnn.gz`, defines the _name_ of the resource, which is a shorthand you can
```

use in parameters without having to specify the full URL (in brackets after the
name).

The second line of each entry contains a short description of the resource.

## Installing known resources

You can install resources with the `ocrd resmgr download` command. It expects
the name of the processor as the first argument and either the name or URL of a
resource as a second argument.

> Likewise, model distribution is not currently centralised within OCR-D though we

Reviewer: ```suggestion
Although model distribution is not currently centralised within OCR-D, we
```

are working towards a central model repository.

For example, to install the `LatinHist.pyrnn.gz` resource for `ocrd-cis-ocropy-recognize`:

```
ocrd resmgr download ocrd-cis-ocropy-recognize LatinHist.pyrnn.gz
# or
ocrd resmgr download ocrd-cis-ocropy-recognize https://github.com/chreul/OCR_Testdata_EarlyPrintedBooks/raw/master/LatinHist-98000.pyrnn.gz
```

This will look up the resource in the [bundled resource and user databases](#user-database), download,
unarchive (where applicable) and store it in the [proper location](#where-is-the-data).

**NOTE:** The special name `*` can be used instead of a resource name/url to
download *all* known resources for this processor. To download all tesseract models:

```sh
ocrd resmgr download ocrd-tesserocr-recognize '*'
```

> (Note that `*` must be in quotes or escaped because of shell wildcard expansion)

Reviewer: ```suggestion
(Note that `*` must be in quotes or escaped to avoid wildcard expansion in the shell.)
```

## Installing unknown resources

> If you need to install a resource that OCR-D doesn't know of, than can be achieved with the `--any-url/-n` flag to `ocrd resmgr download`:

Reviewer: ```suggestion
If you need to install a resource which OCR-D doesn't know of, that can be achieved by passings its URL in combination with the `--any-url/-n` flag to `ocrd resmgr download`:
```

To install a model for `ocrd-tesserocr-recognize` that is located at `https://my-server/mymodel.traineddata`.

```
ocrd resmgr download -n ocrd-tesserocr-recognize https://my-server/mymodel.traineddata
```

This will download and store the resource in the [proper location](#where-is-the-data) and create a stub entry in the
[user database](#user-database).  You can then use it as the parameter value for the `model` parameter:

```
ocrd-tesserocr-recognize -P model mymodel
```

## List installed resources

> The `ocrd resmgr list-installed` command has the same output format as `ocrd resmgr list-available` but instead

Reviewer: ```suggestion
The `ocrd resmgr list-installed` command has the same output format as `ocrd resmgr list-available`. But instead
```

ocrd-cis-ocropy-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-OCRO -P model fraktur-jze.pyrnn.gz
```

## Calamari / ocrd_calamari

Calamari models are Tensorflow model directories. For distribution, this
directory is usually packed to a tarball or ZIP file. Once downloaded, these
containers must be unpacked to a directory again. `ocrd resmgr` handles this
for you, so you just need the name of the resource in the database.

The Calamari-OCR project also maintains a [repository of models](https://github.com/Calamari-OCR/calamari_models).

To use a specific model with OCR-D's calamari wrapper
[ocrd_calamari](https://github.com/OCR-D/ocrd_calamari) and more specifically,
the `ocrd-calamari-recognize` processor, use the `checkpoint_dir` parameter:

```sh
# To use the "default" model, i.e. the one trained on GT4HistOCR by QURATOR
ocrd-calamari-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-CALA
# To use your own trained model
ocrd-calamari-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-CALA -P checkpoint_dir /path/to/modeldir
# or, to be able to control which checkpoints to use:
ocrd-calamari-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-CALA -P checkpoint '/path/to/modeldir/*.ckpt.json'

> ```

Reviewer 2: this means I don't have to specify any model parameter if I want to use the default model. But if I want to use a different model downloaded by `resmgr` , I still have to specify the path?

Author: If the resource was downloaded by `resmgr`, you don't need to specify the full path, just the name of the resource. But if you downloaded a model to some other place that [`resmgr` does not know about](https://ocr-d.de/en/spec/ocrd_tool#file-parameters), you must still specify the full path.

name).

The second line of each entry contains a short description of the resource.

## Installing known resources

You can install resources with the `ocrd resmgr download` command. It expects
the name of the processor as the first argument and either the name or URL of a
resource as a second argument.

Although model distribution is not currently centralised within OCR-D, we
are working towards a central model repository.

For example, to install the `LatinHist.pyrnn.gz` resource for `ocrd-cis-ocropy-recognize`:

```
ocrd resmgr download ocrd-cis-ocropy-recognize LatinHist.pyrnn.gz
# or
ocrd resmgr download ocrd-cis-ocropy-recognize https://github.com/chreul/OCR_Testdata_EarlyPrintedBooks/raw/master/LatinHist-98000.pyrnn.gz
```

This will look up the resource in the [bundled resource and user databases](#user-database), download,
unarchive (where applicable) and store it in the [proper location](#where-is-the-data).

**NOTE:** The special name `*` can be used instead of a resource name/url to
download *all* known resources for this processor. To download all tesseract models:

```sh
ocrd resmgr download ocrd-tesserocr-recognize '*'
```

> (Note that `*` must be in quotes or escaped to avoid wildcard expansion in the shell.)

Reviewer: ```suggestion
**NOTE:** Equally, the special processor `*` can be used instead of a processor and a resource
to download *all* known resources for *all* installed processors:

    ocrd resmgr download '*'

(In either case, `*` must be in quotes or escaped to avoid wildcard expansion by the shell.)
```

OCR(-D) community, please tell us in the [OCR-D gitter
chat](https://gitter.im/OCR-D/Lobby) so we can add it to the database.

## Where is the data

The lookup algorithm is [defined in our specifications](https://ocr-d.de/en/spec/ocrd_tool#file-parameters)

In order of preference, a resource `<name>` for a processor `ocrd-foo` is searched at:

* `$VIRTUAL_ENV/share/ocrd-resources/ocrd-foo/<name>`
* `$HOME/.config/ocrd-resources/ocrd-foo/<name>`
* `$HOME/.local/share/ocrd-resources/ocrd-foo/<name>`
* `$HOME/.cache/ocrd-resources/ocrd-foo/<name>`
* `$PWD/ocrd-resources/ocrd-foo/<name>`

We recommend using the `$VIRTUAL_ENV` location, which is also the default. But
you can override the location to store data with the `--location` option, which can
be `cwd`, `virtualenv`, `config`, `data` and `cache` resp.

```sh
# will download to $PWD/ocrd-resources/ocrd-anybaseocr-dewarp/latest_net_G.pth
ocrd resmgr download --location cwd ocrd-anybaseocr-dewarp latest_net_G.pth
# will download to $HOME.cache/ocrd-resources/ocrd-anybaseocr-dewarp/latest_net_G.pth
ocrd resmgr download --location cache ocrd-anybaseocr-dewarp latest_net_G.pth
```

## Changing the default resource directory

> The `$VIRTUAL_ENV` default location is reasonable because we heavily advertise
> using virtual environments and is compatible with
> [ocrd_all](https://github.com/OCR-D/ocrd_all).
> 
> However, there are use cases where the `config`/`data/`/`cache` or even the
> `cwd` option should be the default (or only) location to store resources and
> resolve file parameters.

Reviewer: ```suggestion
The `$XDG_DATA_HOME` default location is reasonable because
models are usually large files which should persist across different deployments,
both native and containerized, both single-module and [ocrd_all](https://github.com/OCR-D/ocrd_all).
Moreover, that variable can easily be overridden during installation.

However, there are use cases where `system` or even `cwd` should be
used as location to store resources, hence the `--location` option.
```

only be stored in a single location. If the default location (`virtualenv`) is
not the place you want to use for tesseract models, consider [changing the default location
in the OCR-D config file](#changing-the-default-resource-directory).

**NOTE:** For reasons of effiency and to avoid duplicate models, all `ocrd-tesserocr-*` processors
reuse the resource directory for `ocrd-tesserocr-recognize`.

If the `TESSDATA_PREFIX` environemnt variable is set when any of the tesseract processors
are called, it will be the location to look for resources instead of the default.

OCR-D's Tesseract wrapper,
[ocrd_tesserocr](https://github.com/OCR-D/ocrd_tesserocr) and more
specifically, the `ocrd-tesserocr-recognize` processor, expects the name of the
model(s) to be provided as the `model` parameter. Multiple models can be
combined by concatenating with `+` (which generally improves accuracy but always slows processing):

```sh
# Use the deu and frk models
ocrd-tesserocr-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-TESS -P model 'deut+frk'
# Use the Fraktur model
ocrd-tesserocr-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-TESS -P Fraktur
```

# Models and docker

> We recommend a two-step process to make models available in Docker. First
> download all the models that you want to use on the host system. When running
> the docker container, mount that local directory into the container alongside
> the data you want to process.
> 
> Download the models to `$HOME/.local/share/ocrd-resources`:
> 
> ```sh
> ocrd resmgr download --location data ocrd-tesserocr-recognize eng.traineddata
> ocrd resmgr download --location data ocrd-calamari-recognize default
> # ...
> ```
> 
> Run the `ocrd_all` Docker container:
> 
> ```sh
> docker run --user $(id -u) --workdir /data \
>   --volume $PWD:/data \
>   --volume $HOME/.local/cache/ocrd-resources:/ocrd-resources \
>   ocrd_all ocrd-tesserocr-recognize -I IN -O OUT -P model eng
> ```

Reviewer: ```suggestion
We recommend keeping all downloaded resources in a persistent host directory, 
separate of the `ocrd/all` Docker container and data directory, and mounting that
resource directory into a specific path in the container alongside the data directory.
The host resource directory can be empty initially. Each time you run the Docker container,
your processors will access the host directory to resolve resources, and you can download
additional models into that location using `ocrd resmgr`.

The following will assume (without loss of generality) that your host-side data path is under `./data`, and the host-side resource path is under `./models`:

- To download models to `/usr/share/ocrd-resources` in Docker and `./models` in the host FS, use `virtualenv` as storage location:

        docker run --user $(id -u) --workdir /data \
          --volume $PWD/data:/data \
          --volume $PWD/models:/usr/share/ocrd-resources \
        ocrd/all \
        ocrd resmgr download --location virtualenv ocrd-tesserocr-recognize eng.traineddata\; \
        ocrd resmgr download --location virtualenv ocrd-calamari-recognize default\; \
        ...

- To run processors, as usual do:

        docker run --user $(id -u) --workdir /data \
          --volume $PWD/data:/data \
          --volume $PWD/models:/usr/share/ocrd-resources \
          ocrd/all ocrd-tesserocr-recognize -I IN -O OUT -P model eng

```

Reviewer: And IIRC for that very reason we also wanted to make `--location virtualenv` the _default_ in the Docker build...

Reviewer: > > And IIRC for that very reason we also wanted to make `--location virtualenv` the _default_ in the Docker build...
> 
> This works on the top-level venv but what about the sub-venvs? They have their own `$VIRTUAL_ENV` and hence the `ocrd resmgr` in the sub-venvs will not see the model installed to top-level `$VIRTUAL_ENV/share/ocrd-resources` :( Correct me if I'm wrong but I think we do need either XDG or re-introduce an absolute path like `/usr/share/local/ocrd-resources` ...

Uh-oh, yes, you are right! But it's not `ocrd resmgr` which is the problem – since we would only need that at the top-level venv – but the processors' `resource_manager` search path. So `virtualenv` is out of question not only for Docker, but also for native ocrd_all deployments, right? Therefore we must always use one of the XDG paths, except maybe for users who install modules individually. But the latter can still live with XDG I guess, so I recommend removing `virtualenv` **entirely** from the implementation and spec.

Do you want me to update the above suggestion in terms of XDG `data`?

Author: > So virtualenv is out of question not only for Docker, but also for native ocrd_all deployments, right?

Afraid so :( I'll adapt the spec and implementation to default to `data` for consistency.

I am also not happy about `$HOME` being `/` or `/root` depending on the `docker run` call. Reintroduce /usr/local/share (shorthand `system`)?

> Do you want me to update the above suggestion in terms of XDG data?

Yes, please, thanks.

Reviewer: > > So virtualenv is out of question not only for Docker, but also for native ocrd_all deployments, right?
> 
> Afraid so :( I'll adapt the spec and implementation to default to `data` for consistency.

You mean `and default to`, right? (We should rid of `virtualenv` as a possibility itself.)

> I am also not happy about `$HOME` being `/` or `/root` depending on the `docker run` call. Reintroduce /usr/local/share (shorthand `system`)?

Yes, we don't want to get into that mess. We could from now on set `ENV HOME=/` in our Dockerfiles (ocrd/all as well as ocrd/tesserocr etc), so `XDG_DATA_HOME` would be `/.local/share`, or even set `ENV XDG_DATA_HOME=/usr/local/share` and `ENV XDG_CONFIG_HOME=/usr/local/etc` and so on. 

Regarding absolute paths like `/usr/local/share/ocrd-resources` – agreed, but only if configurable (not fixed). So if `XDG_CONFIG_HOME` can be relied upon (even in Docker, somehow), then core can always find its `ocrd/config.yml` (the same for top-level and sub venvs), look up the default `resouce_location` (from the enum `data` / `cache` / `config` or a string for the absolute path), and be done. 

That one assumption (`XDG_CONFIG_HOME`) _must_ be satisfiable – after all, what's the point of configuring paths if you cannot even reliably find the path to that config?

Reviewer: In light of the recent discussion, with `virtualenv` disqualifying itself due to our sub-venv mechanics, and the need to have at least one stable entry point for configuration, I wonder what's the point of keeping the many-valued `resource_location` logic and a config file for it at all. The deployment just needs to make sure `XDG_DATA_HOME` is set sensibly (for `$XDG_DATA_HOME/ocrd-resources`), everything else is unnecessary complexity.

Reviewer: BTW: considering that https://github.com/OCR-D/ocrd_all/pull/234 simply delegates to `ocrd resmgr`, we could even explain to just do

    docker run --user $(id -u) --volume $PWD/models:/usr/local/share/ocrd-resources \
        ocrd/all make -C /build install-models

Do you want me to include even that in the above suggestion?

Author: > BTW: considering that [OCR-D/ocrd_all#234](https://github.com/OCR-D/ocrd_all/pull/234) simply delegates to `ocrd resmgr`, we could even explain to just do
> 
> ```
> docker run --user $(id -u) --volume $PWD/models:/usr/local/share/ocrd-resources \
>     ocrd/all make -C /build install-models
> ```
> 
> Do you want me to include even that in the above suggestion?

I love that this is possible but I do think that we should familiarize users with `ocrd resmgr`, whereas `make install-models` was just a stop-gap solution.

Reviewer: > I love that this is possible but I do think that we should familiarize users with `ocrd resmgr`, whereas `make install-models` was just a stop-gap solution.

Okay how about a `ocrd resmgr download '*'` then?

In contrast to `install-models`, this leaves the problem of knowing which processors/modules are actually installed to the user (or to `ocrd resmgr`).

Author: > EDIT: What should happen if I `ocrd resmgr download ocrd-kraken-recognize '*'` although I don't have ocrd_kraken installed

It will install the models just fine. You could also do `ocrd resmgr download ocrd-foo /vmlinuz` and it would copy `/vmlinuz` to `$HOME/.local/share/ocrd-foo/vmlinuz`, no questions asked.

> In contrast to install-models, this leaves the problem of knowing which processors/modules are actually installed to the user (or to ocrd resmgr), though.

I also saw your proposal in ocrd_all but, while an elegant use of make, it artificially restricts what can be done. From experience, there probably are unexpected use cases, where that restriction would require users to hack around those rules.

> Okay how about a ocrd resmgr download '*' then?

What we could do is: Find all `ocrd-*` in `$PATH` and use that as the set of processors for which to download resources. Would that achieve what you want?

Reviewer: > > What should happen if I `ocrd resmgr download ocrd-kraken-recognize '*'` although I don't have ocrd_kraken installed
> 
> It will install the models just fine. You could also do `ocrd resmgr download ocrd-foo /vmlinuz` and it would copy `/vmlinuz` to `$HOME/.local/share/ocrd-foo/vmlinuz`, no questions asked.

IMHO we should at least warn that there's no `ocrd-foo` here. (Maybe it was misspelled, or the user forgot to install it?)

> I also saw your proposal in ocrd_all but, while an elegant use of make, it artificially restricts what can be done. From experience, there probably are unexpected use cases, where that restriction would require users to hack around those rules.

If you are arguing against [encapsulation in ocrd_all](https://github.com/OCR-D/ocrd_all/pull/234#pullrequestreview-574250990), then I must respond that just as `OCRD_MODULES` can be changed freely to extend or reduce the set of modules to install, here `OCRD_MODULES` would be used for the set of models to install – it's what should be expected for `make install-models` per analogy.

But if all you wanted to say was that `ocrd resmgr download '*'` should not suffer any restrictions, then I tend to agree with you. However, remember we still want to aggregate the `resource_list.yml` during ocrd_all's `make all` in the long term. So naturally the list of available models would derive from the list of installed modules. The connection is already there! If the user installs other modules manually, fine, they will have only unregistered resources. Still, `*` should give only the registered models (i.e. from modules included in `make all`).

Author: > IMHO we should at least warn that there's no ocrd-foo here. (Maybe it was misspelled, or the user forgot to install it?)

Agreed, 22fb2c62. 

> If you are arguing against encapsulation in ocrd_all, then I must respond that just as OCRD_MODULES can be changed freely to extend or reduce the set of modules to install, here OCRD_MODULES would be used for the set of models to install – it's what should be expected for make install-models per analogy.

Point taken, lets continue discussing at https://github.com/OCR-D/ocrd_all/pull/234#pullrequestreview-574250990

Author: Replaced the section with your proposed text in f76c0a7b

BTW `ocrd resmgr download '*'` is now implemented in OCR-D/core#559.

Reviewer: > BTW `ocrd resmgr download '*'` is now implemented in [OCR-D/core#559](https://github.com/OCR-D/core/pull/559).

Fantastic!
